File: Java/Data Structures & Algorithms Module - Java/Week10/BinarySearchTree.java
/*
BinarySearchTree.java
Billy Cussen
24/11/2020 - Updated on 25/11/2020
*/
public class BinarySearchTree<T extends Comparable<T>> implements BinarySearchTreeInterface<T>{

    BSTNode<T> root;

    @Override
    public void insert(T elem){
        if(isEmpty()){
            this.root = new BSTNode<T>(elem);
        } else{
            insertHelper(elem, this.root);
        }
    }

    private void insertHelper(T elem, BSTNode<T> current){
        if(elem.compareTo(current.element) == -1){
            if(current.left==null){
                current.left = new BSTNode<T>(elem);
            } else{
                insertHelper(elem,current.left);
            }
        } else{
            if(current.right==null){
                current.right = new BSTNode<T>(elem);
            } else{
                insertHelper(elem,current.right);
            }
        }
    }

    @Override
    public void inOrder(){
        if(isEmpty()){
            throw new NullPointerException("Can't do in order traversal, the BST is Empty!");
        } else{
            System.out.println("***BINARY SEARCH TREE IN ORDER TRAVERSAL***");
            inOrderHelper(this.root);
            System.out.println("***END***");
        }
    }

    private void inOrderHelper(BSTNode<T> current){
        if(current == null){
            return;
        } else{
            inOrderHelper(current.left);
            System.out.println(current);
            inOrderHelper(current.right);
        }
    }

    @Override
    public boolean isEmpty(){
        return this.root == null;
    }

    @Override
    public int size(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Size, BST is Empty!");
        } else{
            return sizeHelper(this.root);
        }
    }

    private int sizeHelper(BSTNode<T> current){
        if(current==null){
            return 0;
        } 
        return 1 + sizeHelper(current.left)+sizeHelper(current.right);
    }

    @Override
    public T getMin(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Min, BST is Empty!");
        }
        return getMinHelper(this.root);
    }

    private T getMinHelper(BSTNode<T> current){
        if(current.left==null){
            return current.element;
        }
        return getMinHelper(current.left);
    }

    @Override
    public T getMax(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Max, BST is Empty!");
        }
        return getMaxHelper(this.root);
    }

    private T getMaxHelper(BSTNode<T> current){
        if(current.right == null){
            return current.element;
        }
        return getMaxHelper(current.right);
    }

    @Override
    public int getDepth(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Depth, BST is Empty!");
        }
        return getDepthHelper(this.root);
    }

    private int getDepthHelper(BSTNode<T> current){
        if(current==null){
            return 0;
        }

        int left = getDepthHelper(current.left);
        int right = getDepthHelper(current.right);

        if(left>right){
            return 1 + left;
        } else{
            return 1 + right;
        }
    }

    @Override
    public T remove(T element){
        return null;
    }

    @Override
    public boolean contains(T element){
        if(isEmpty()){
            throw new NullPointerException("Can't locate "+element+", the BST is empty!");
        } 
        return containsHelper(this.root,element);
    }

    private boolean containsHelper(BSTNode<T> current, T element){
        if(current == null){
            return false;
        }

        if(element.compareTo(current.element) == 0){
            return true;
        } else if (element.compareTo(current.element) == -1){
            return containsHelper(current.left, element);
        } else{
            return containsHelper(current.right,element);
        }
    }

}

File: Python/Data Structures & Algorithms Module - Python/Week10/BinarySearchTree.py
"""
BinarySearchTree.py
Billy Cussen
25/11/2020
"""

from BSTNode import BSTNode

class BinarySearchTree:

    def __init__(self):
        self.root = None

    def insert(self, element):
        if self.isEmpty():
            self.root = BSTNode(element)
        else:
            self.insertHelper(element,self.root)

    def insertHelper(self, element, current):
        if ((element > current.element) - (element < current.element)) == -1:
            if(current.left == None):
                current.left = BSTNode(element)
            else:
                self.insertHelper(element,current.left)
        elif ((element > current.element) - (element < current.element)) == 1:
            if(current.right == None):
                current.right = BSTNode(element)
            else:
                self.insertHelper(element,current.right)

    def inOrder(self):
        if self.isEmpty():
            raise Exception("Can't do in order traversal, the BST is empty!")
        print("***BST In Order Traversal***\n")
        self.inOrderHelper(self.root)
        print("***END***")

    def inOrderHelper(self, current):
        if current == None:
            return
        
        self.inOrderHelper(current.left)
        print(current.toString())
        self.inOrderHelper(current.right)

    def isEmpty(self):
        return self.root == None

    def size(self):
        if self.isEmpty():
            raise Exception("Can't get Size, BST is Empty!")
        return self.sizeHelper(self.root)

    def sizeHelper(self, current):
        if current == None:
            return 0
        return 1 + self.sizeHelper(current.left)+self.sizeHelper(current.right)

    def getMin(self):
        if self.isEmpty():
            raise Exception("Can't get Minimum, BST is empty!")
        return self.getMinHelper(self.root)

    def getMinHelper(self,current):
        if current.left == None:
            return current.element
        else:
            return self.getMinHelper(current.left)   

    def getMax(self):
        if self.isEmpty():
            raise Exception("Can't get Max, BST is Empty!") 
        return self.getMaxHelper(self.root)

    def getMaxHelper(self, current):
        if(current.right==None):
            return current.element
        else:
            return self.getMaxHelper(current.right)

    def getNode(self, element):
        if self.isEmpty():
            raise Exception("Can't get Node, BST is Empty!")
        return self.getNodeHelper(self.root,element)

    def getNodeHelper(self, current, element):
        if(current==None):
            return None
        if ((element > current.element)-(element < current.element)) == 0:
            return current.toString()
        elif ((element > current.element)-(element < current.element)) == -1:
            return self.getNodeHelper(current.left,element)
        else:
            return self.getNodeHelper(current.right,element)

    def getParent(self, element):
        if self.isEmpty():
            raise Exception("Unable to get Parent, BST is Empty!")
        elif ((element > self.root.element)-(element < self.root.element))==0:
            raise Exception("Unable to get Parent, this element is the Root!")
        return self.getParentHelper(self.root,element)

    def getParentHelper(self, current, element):
        if(current==None):
            return None

        if ((element > current.element)-(element < current.element)) == -1:
            if((element > current.left.element)-(element < current.left.element)==0):
                return current.toString()
            else:
                return self.getParentHelper(current.left, element)
        else:
            if((element > current.right.element) - (element < current.right.element)==0):
                return current.toString()
            else:
                return self.getParentHelper(current.right, element)